### PR TITLE
Provide example of writing a concrete service

### PR DESCRIPTION
**Pull request by drbenmorgan**
As requested by @pfranchini, provide an example of implementing a custom service (use case here will be dead cells). To cover:

- Basic interface
- Loading//configuring from flreconstruct scripts

May need updates to `flreconstruct` implementation so script allows starting/loading services. Should roll that in here so things are consistent.

### CODE REVIEW

**Review by pfranchini — DISMISSED** [review]

**Comment by cherylepatrick** [comment]
Hi @drbenmorgan and thanks for looking at this. @pfranchini and I are planning to meet next week and attempt to make a service to kill dead channels. Do you have any hints for us newbies?

Do we just add it to the CMakeLists as you did with the plugin service example here, and then copy those other files? Best idea to run it and see what it prints out? (Can you remind us how to include it in our flreconstruct config script?)

Once we've done that, do we add our code to the `main()` function? Or to `builtin_impl()`? Or somewhere else?

And can you give us any clues where we should look to see how to access stuff like the hit data, and how to write back to a brio file?

Thanks in advance!

**Comment by pfranchini** [comment]
Since I could not make another attempt I was working on to work, I went back to the code example Ben has provided, which should be
`source/falaise/test/test_service_plugin.cxx`

With few changes I have compiled it. Now, how can I load it into a Falaise module?

**Comment by drbenmorgan** [comment]
The "canonical" way is roughly as follows (some code omitted for brevity):

```c++
#include "my_service.h"
#include "falaise/services/service_handle.h"

class my_module {
 public:
  my_module() = default;

  my_module(const falaise::property_set& ps, datatools::service_manager& services);

 private:
  snemo::service_handle<my_service> mySVC_;
};

my_module::my_module(const falaise::property_set& ps, datatools::service_manager& services)
    : mySVC_{services} {}
```

The `mySVC_` is a smart pointer to the instance of your service, so can be used to call any member functions of it in the normal way.

**Comment by pfranchini** [comment]
Hi @drbenmorgan, I am using the Deadcell module to play with the service. Is possible to initialize the service member in any other function of the service?

**Comment by drbenmorgan** [comment]
Hi @pfranchini, the initialization (i.e. construction) should always happen in the constructor of the module, but you can call any member function of the service in any other module member function.

**Comment by pfranchini** [comment]
I have built the service in Falaise in `build/Falaise.ben.build/lib64/libplugin_service.so` and the module to test it with
`cmake -DCMAKE_PREFIX_PATH=build/Falaise.ben.build ../SuperNEMO-DeadCellsModule`

but I get

```
nm --undefined-only build/SuperNEMO-DeadCellsModule.build/libDeadCellsModule.so  | grep plugin_service
                 U _ZTI14plugin_service
```
Do you see anything I have obviously missed? Thanks!

**Comment by drbenmorgan** [comment]
Hmm, is the code available anywhere? You could try linking the module to the service library, but I think the issue is more that the service may not be linked into Falaise.

**Comment by pfranchini** [comment]
Fair question ;) The bare-bone of the code is in here https://github.com/pfranchini/Falaise/tree/implement-service-example/source/falaise/snemo/services (forked from yours)
basically plugin_service.cxx/h

I might have messed up the cmake... Thanks

**Comment by drbenmorgan** [comment]
Hi @pfranchini, I think it's more to do with having a `main` function in the library (`plugin_service.cxx`)! I'm going to try and fix this properly by end of tomorrow, as the "plugin_service" needs some work.

File: documentation/flreconstruct/WritingFLReconstructServices.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  return true;
+}
+
+std::string hello_world_svc::say_hello(const std::string& to_who) const {
+  // Implement the service implementation in one of more functions
+  return std::string("Hello " + to_who);
+}
+
+}  // namespace snemo
+```
```
Comment (drbenmorgan) [inline documentation/flreconstruct/WritingFLReconstructServices.md]:
```suggestion
~~~~~
```

**Comment by drbenmorgan** [comment]
Hi @pfranchini, this is now updated with a slightly, I hope, clearer example. I've removed the "plugin_service" as that's not going to work with some detailed work in Falaise/Bayeux. The documentation's also been added to walk through how to add and integrate a service in Falaise, so let me know if that's clear enough.

You should be able to take the `hello_world.h` and `.cc` files as a template for a dead cells service `.h` and `.cc` files in the same location, and add them to the build as per the documentation. I suggest we start that in a separate PR if you're o.k. with the code/docs as updated.

**Comment by pfranchini** [comment]
Thanks very much with that. Yes we can move to another PR.

**Comment by drbenmorgan** [comment]
O.k., if you can approve this one, I'll merge it. You can can then start a branch+PR for the dead cells and we can work up the implementation there!

**Review by pfranchini — APPROVED** [review]
Thanks Ben!

**Comment by drbenmorgan** [comment]
Thanks @pfranchini, merged to develop, so fetch the changes and rebase your local clone, then should be ready to start the new branch!

**Pull request merged**